storage/statedb/hasher.go

```diff
@@ -148,6 +187,58 @@ func (h *hasher) hashChildren(original node, db *Database) (node, node) {
 	}
 }
 
+type hashResult struct {
+	index     int
+	collapsed node
+	cached    node
+}
+
+func (h *hasher) hashChildrenFromRoot(original node, db *Database, onleaf LeafCallback) (node, node) {
+	switch n := original.(type) {
+	case *shortNode:
+		// Hash the short node's child, caching the newly hashed subtree
+		collapsed, cached := n.copy(), n.copy()
+		collapsed.Key = hexToCompact(n.Key)
+		cached.Key = common.CopyBytes(n.Key)
+
+		if _, ok := n.Val.(valueNode); !ok {
+			collapsed.Val, cached.Val = h.hash(n.Val, db, false)
+		}
+		return collapsed, cached
+
+	case *fullNode:
+		// Hash the full node's children, caching the newly hashed subtrees
+		collapsed, cached := n.copy(), n.copy()
+
+		hashResultCh := make(chan hashResult, 16)
+		numRootChildren := 0
+		for i := 0; i < 16; i++ {
+			if n.Children[i] != nil {
+				numRootChildren++
+				go func(i int, n node) {
+					childHasher := newHasher(onleaf)
```

Comment:
If you use `h.onleaf` instead of `onleaf`, you don't have to take `onleaf` as an argument.
```suggestion
					childHasher := newHasher(h.onleaf)
```